controllers/actions.summerwind.net/horizontal_runner_autoscaler_batch_scale.go

```diff
@@ -157,72 +200,40 @@ func (s *batchScaler) batchScale(ctx context.Context, batch batchScaleOperation)
 		} else if amount < 0 {
-			var reservations []v1alpha1.CapacityReservation
-
-			var (
-				found    bool
-				foundIdx int
-			)
-
-			for i, r := range copy.Spec.CapacityReservations {
-				r := r
-				if !found && r.Replicas+amount == 0 {
-					found = true
-					foundIdx = i
-				} else {
-					// Note that we nil-check max replicas because this "fix" is needed only when there is the upper limit of runners.
-					// In other words, you don't need to reset effective time and expiration time when there is no max replicas.
-					// That's because the desired replicas would already contain the reservation since it's creation.
-					if found && copy.Spec.MaxReplicas != nil && i > foundIdx+*copy.Spec.MaxReplicas {
-						// Update newer CapacityReservations' time to now to trigger reconcile
-						// Without this, we might stuck in minReplicas unnecessarily long.
-						// That is, we might not scale up after an ephemeral runner has been deleted
-						// until a new scale up, all runners finish, or after DefaultRunnerPodRecreationDelayAfterWebhookScale
-						// See https://github.com/actions/actions-runner-controller/issues/2254 for more context.
-						r.EffectiveTime = metav1.Time{Time: now}
-
-						// We also reset the scale trigger expiration time, so that you don't need to tweak
-						// scale trigger duratoin depending on maxReplicas.
-						// A detailed explanation follows.
-						//
-						// Let's say maxReplicas=3 and the workflow job of status=canceled result in deleting the first capacity reservation hence i=0.
-						// We are interested in at least four reservations and runners:
-						//   i=0   - already included in the current desired replicas, but just got deleted
-						//   i=1-2 - already included in the current desired replicas
-						//   i=3   - not yet included in the current desired replicas, might have been expired while waiting in the queue
-						//
-						// i=3 is especially important here- If we didn't reset the expiration time of 3rd reservation,
-						// it might expire before a corresponding runner is created, due to the delay between the expiration timer starts and the runner is created.
-						//
-						// Why is there such delay? Because ARC implements the scale duration and expiration as such...
-						// The expiration timer starts when the reservation is created, while the runner is created only after the corresponding reservation fits within maxReplicas.
-						//
-						// We address that, by resetting the expiration time for fourth(i=3 in the above example) and subsequent reservations when the first reservation gets cancelled.
-						r.ExpirationTime = metav1.Time{Time: now.Add(scale.trigger.Duration.Duration)}
-					}
-
-					reservations = append(reservations, r)
-				}
+			// Remove the requested number of reservations unless there are not that many left
+			if len(copy.Spec.CapacityReservations) > -amount {
+				copy.Spec.CapacityReservations = copy.Spec.CapacityReservations[-amount:]
```

Comment:
You changed it from FIFO to FILO, right? If so, I guess that's incorrect- the earlier workflow jobs would correspond to earlier capacity reservations. Every scale down request(that corresponds to `completed` or `cancelled` workflow job webhook event) should thus result in removing the earliest capacity reservations that matches. As the `amount` is either 1 or -1, foundIdx should  be 0, resulting in removing the first capacity reservation on every scale down.

Reply:
No, it's still FIFO. The first in is i = 0, next in is i = 1, this code removes i = 0. 

This code has the added advantage of working properly when `amount` is a number other than +1 or -1.

Reply:
> When amount == -1 this code simply removes the entry at index zero:

Got it! I somehow missed the preceding `-`.

> The code in this PR has the added advantage of working properly when amount is a number other than +1 or -1.

Why do we need this...? (Just unsure

If we are to support other amounts, it would be very likely we add "amount" also to the capacity reservation, so that we no longer need multiple capacity reservations to add multiple replicas- instead we need only one of those at a time.

Some context: Currently, a scale trigger is a "queued" capacity reservation. So a scale trigger's "amount" corresponds to how many replicas are added/removed via a corresponding capacity reservation. The amount if currently either 1 or -1 because capacity reservation can only express a single replica(it has no `amount` field).

Reply:
Well anyway, your goal for this pull request was two-fold, right?

- First, you wanted to let ARC update capacity reservation expiration on both scale up and scale down attempt, rather than only on scale down attempt that we currently do
- Second, you wanted to move the expiration update logic before the `getValidCapacityReservations`, so that (hopefully) capacity reservation can start to expire only after it was added a real replica.

Your theory is that these two enable us to (hopefully) fix the issue that a capacity reservation can expire while in the queue(due to it being above maxReplicas), right?

Reply:
I've updated the pull request title based on my understanding. WDYT?

Reply:
@mumoshu wrote
> If we are to support other amounts, it would be very likely we add "amount" also to the capacity reservation, so that we no longer need multiple capacity reservations to add multiple replicas- instead we need only one of those at a time.

Capacity reservations already have an "amount" field, it's just called `Replicas`. The problem is that if you have 1 capacity reservation, and it is reserving, for example, 3 replicas, how do you scale down by 1? You have to decrement the Replica field in the reservation rather than just remove it. Also, in this scenario, calculating the number of reserved replicas mean looping over all the reservations and summing their Replica fields, rather than just taking the length of the replica array. 

Supporting reservations reserving other than exactly 1 replica makes all of the code more complex and IMHO it is not worth it. This PR ensures that every capacity reservation reserves exactly 1 replica, no matter what the scale up amount is. 

> If we don't, amount would never be other than 1 or -1- so we don't need support for that in code.

Unfortunately, `scaleOps` have a `scale.trigger.Amount`, and the old code read that field and used it (somewhat incorrectly). If the field is there, someone is likely to use it, and this PR ensures that when they do, the scaling will still work.  I had a choice of either adding an assertion that the value was either 1 or -1 or fixing the code to handle other values, and I chose to fix it. 

> I've updated the pull request title based on my understanding. WDYT?
 
I don't like "Update unconsumed HRA capacity reservation's expiration more frequently" because that implies that the only problem this fixes is somehow related to the update frequency. One problem this PR fixes is that under certain conditions (some rare, some common), a reservation that was backlogged would not have its expiration time adjusted at all. 

I would prefer the title "Update all unconsumed HRA capacity reservation's expiration correctly".

Reply:
@Nuru Hey! Thanks for your time.

> "Update all unconsumed HRA capacity reservation's expiration correctly".

Can we somehow rephrase "correctly" to a more concrete alternative?
What do you think you did to make it work "correctly"? My impression was that you tried to make it more correct by changing logic to let ARC update expirations "more frequently", but you have another opinion- right?

Reply:
> Capacity reservations already have an "amount" field, it's just called Replicas. The problem is that if you have 1 capacity reservation, and it is reserving, for example, 3 replicas, how do you scale down by 1? You have to decrement the Replica field in the reservation rather than just remove it

Yeah good point! And that's why I made it only have two values, either 1 or -1. I thought you were willing to do that by saying "The code in this PR has the added advantage of working properly when amount is a number other than +1 or -1."-

> the old code read that field and used it (somewhat incorrectly). If the field is there, someone is likely to use it, and this PR ensures that when they do, the scaling will still work. I had a choice of either adding an assertion that the value was either 1 or -1 or fixing the code to handle other values, and I chose to fix it.

Okay! So- you did it just for consistency with existing node, not to actually add support for non-zero amount. Am I following you correctly?

Reply:
> Your theory is that these two enable us to (hopefully) fix the issue that a capacity reservation can expire while in the queue(due to it being above maxReplicas), right?

@Nuru Did this part make sense to you? I'm still missing the exact goal of your seemingly awesome work. I do believe you put a lot of considerations and efforts into this work but that blurred the goal of this pull request to me.

Re:

> "Update all unconsumed HRA capacity reservation's expiration correctly".

I think this is good to go once we can agree on the definition of "correct" here. My impression was it was working as expected and has been correct already. However you thought it incorrect and therefore phrased "correctly". All in all, the term "correct" is more subjective. I'm more interested in exactly which part of the existing logic looked incorrect to you and how you fixed it!

And my understanding was https://github.com/actions/actions-runner-controller/pull/2502#discussion_r1169376047. Is that correct?

Reply:
Maybe https://github.com/actions/actions-runner-controller/pull/2502/files#r1169468039 relates to the correctness, too?

Reply:
The correctness I am referring to is explained in the PR description:

>The existing code has a number of issues, particularly in the case where capacity reservations exceeded maxReplicas, that may have only been obvious when minReplicas was set to zero, HRA scaleUpTriggers.duration was close to the length of time jobs actually took to run, and more than maxReplicas * 3 jobs were queued in a very short period of time.
>
>Of note is that with foundIdx = 0 (the normal case) and maxReplicas = 3 as posited in the comments, i == 3 would have been skipped because 3 > 0+3 is false.

In other words, in the normal case, the oldest backlogged reservation was not having its expiration time extended. Once it expired, the next in line would become the oldest backlogged reservation and its exipration time would top being extended. So the backlogged items were still being expired prematurely.

Reply:
Thanks a lot for your patience and effort @Nuru!

So maybe you're correct! I still think the PR description contains only one thing fixed in this pull request. I'm still interested in the whole picture of this pull request.

Based on your explanation, I can definitely think that probably "Summary of changes:" in your PR description covers both "real" fixes that I can agree, and other tiny improvements for potential issues(have not been reported yet or not too important).

Presuming you're out of assigned time now, I'll try to sort this out somehow- As you changed quite a lot of code here, probably I'd need to write tests for existing and new code, and group your change descriptions into seemingly-important fixes and nice-to-have fixes. Maybe it will take a week or two. Hope that works for you.

Reply:
@mumoshu I would be grateful if you wrote tests. Thank you.

Please, in particular, test the case where capacity reservations exceed `maxReplicas`, `minReplicas` is set to zero, HRA `scaleUpTriggers.duration` are close to the length of time jobs actually take to run, and more than `maxReplicas * 3 jobs` are queued in a very short period of time. I expect that in that situation, reservations are expired that should not be, leading to the pool scaling to zero with jobs still on the queue. 

Yes, a lot of the PR falls under the category of "code cleanup", fixing trivial bugs or things that work now because of the limitations of how it is being used, but could break in the future with different usage (like scale amount other than 1). 

This can wait a couple of weeks for you to write tests. Please prioritize #2520, which is having a greater impact in production.